documentation/bolt_types_reference.md

```diff
@@ -0,0 +1,319 @@
+
+### `action`
+
+**Returns**
+
+`String`: The type of result (`task`, `command`, etc.)
+
+### `error`
+
+**Returns**
+
+`Error`: An object constructed from the `_error` field of the result's value.
+
+### `message`
+
+**Returns**
+
+`String`: The `_output` field of the result's value.
+
+### `ok`
+
+**Returns**
+
+`Boolean`: Whether the result was successful.
+
+### `status`
+
+**Returns**
+
+`String`: Either `success` if the result was successful or `failure`.
+
+### `target`
+
+**Returns**
+
+`Target`: The target the result is from.
+
+### `value`
+
+**Returns**
+
+`Hash`: The value of the result.
+
+
+## `ResultSet`
+
+A `ResultSet` is returned from an execution function and contains one or more `Result`
+objects. An [apply action](https://puppet.com/docs/bolt/latest/applying_manifest_blocks.html#)
+returns a `ResultSet` with one or more `ApplyResult` objects.
+
+The following functions are available to `ResultSet` objects. They are accessed with
+dot notation `$result_set.function`.
+
+### `[]`
+
+Access a range of results or the _i_ th result in the set.
+
+> **Note:** This function does not use dot notation and is called directly on the
+  `Result`, e.g. `$result_set[0]`, `$result_set[0, 2]`
+
+**Returns**
+
+`Variant[Result, ApplyResult, Array[Variant[Result, ApplyResult]]]`: The accessed results.
+
+### `count`
+
+**Returns**
+
+`Integer`: The number of results in the set.
+
+### `empty`
+
+**Returns**
+
+`Boolean`: Whether the set is empty.
+
+### `error_set`
+
+**Returns**
+
+`ResultSet`: The set of failing results.
+
+### `filter_set(block)`
+
+**Parameters**
+
+- `block`
+
+  The block to filter the set by.
+
+**Returns**
+
+`ResultSet`: The set of filtered results.
+
+### `find(String $target_name)`
+
+**Parameters**
+
+- `String` target_name
+
+  The string name of the target to retrive a result for.
+
+**Returns**
+
+`Variant[Result, ApplyResult]`: The target's result.
+
+### `first`
+
+**Returns**
+
+`Variant[Result, ApplyResult]`: The first result in the set. Useful for unwrapping single results.
+
+### `names`
+
+**Returns**
+
+`Array[String]`: The names of all targets that have results in the set.
+
+### `ok`
+
+**Returns**
+
+`Boolean`: Whether all results were successful. Equivalent to `$result_set.error_set.empty`.
+
+### `ok_set`
+
+**Returns**
+
+`ResultSet`: The set of successful results.
+
+### `results`
+
+**Returns**
+
+`Array[Variant[Result, ApplyResult]]`: All results in the set.
+
+### `targets`
+
+**Returns**
+
+`Array[Target]`: The list of targets that have results in the set.
+
+### `to_data`
+
+**Returns**
+
+`Array[Hash`: An array of serialized representations of each result in the set.
+
+
+## `Target`
+
+The `Target` object represents a target and its specific connection options.
+
+The following functions are available to `ResultSet` objects. They are accessed with
+dot notation `$result_set.function`.
+
+### `config`
+
+**Returns**
+
+`Hash[String, Data]`: The target-level configuration.
```

Comment:
I think including a line about how this is only data from the inventory, though I think it's all inventory data that applies to a target (not just defined specifically for that target)

Reply:
This in addition to the note below.